src/consensus/params.h

```diff
@@ -25,6 +25,77 @@ struct Params {
     int64_t nPowTargetSpacing;
     int64_t nPowTargetTimespan;
     int64_t DifficultyAdjustmentInterval() const { return nPowTargetTimespan / nPowTargetSpacing; }
+
+    /** Maximum block size parameters */
+    uint32_t nMaxSizePreFork;
+    uint64_t nEarliestSizeForkTime;
+    uint32_t nSizeDoubleEpoch;
+    uint64_t nMaxSizeBase;
+    uint8_t nMaxSizeDoublings;
+    int nActivateSizeForkMajority;
+    uint64_t nSizeForkGracePeriod;
+
+    /** Maximum block size of a block with timestamp nBlockTimestamp */
+    uint64_t MaxBlockSize(uint64_t nBlockTimestamp, uint64_t nSizeForkActivationTime) const {
+        if (nBlockTimestamp < nEarliestSizeForkTime || nBlockTimestamp < nSizeForkActivationTime)
+            return nMaxSizePreFork;
+        if (nBlockTimestamp >= nEarliestSizeForkTime + nSizeDoubleEpoch * nMaxSizeDoublings)
+            return nMaxSizeBase << nMaxSizeDoublings;
+
+        // Piecewise-linear-between-doublings growth. Calculated based on a fixed
+        // timestamp and not the activation time so the maximum size is
+        // predictable, and so the activation time can be completely removed in
+        // a future version of this code after the fork is complete.
+        uint64_t timeDelta = nBlockTimestamp - nEarliestSizeForkTime;
+        uint64_t doublings = timeDelta / nSizeDoubleEpoch;
+        uint64_t remain = timeDelta % nSizeDoubleEpoch;
+        uint64_t interpolate = (nMaxSizeBase << doublings) * remain / nSizeDoubleEpoch;
+        uint64_t nMaxSize = (nMaxSizeBase << doublings) + interpolate;
+        return nMaxSize;
+    }
+
+    // Signature-operation-counting is a CPU exhaustion denial-of-service prevention
+    // measure. Prior to the maximum block size fork it was done in two different, ad-hoc,
+    // inaccurate ways.
+    // Post-fork it is done in an accurate way, counting how many ECDSA verify operations
+    // and how many bytes must be hashed to compute signature hashes to validate a block.
+
+    /** Pre-fork consensus rules use an inaccurate method of counting sigops **/
+    uint64_t MaxBlockLegacySigops(uint64_t nBlockTimestamp, uint64_t nSizeForkActivationTime) const {
+        if (nBlockTimestamp < nEarliestSizeForkTime || nBlockTimestamp < nSizeForkActivationTime)
+            return MaxBlockSize(nBlockTimestamp, nSizeForkActivationTime)/50;
+        return std::numeric_limits<uint64_t>::max(); // Post-fork uses accurate method
```

Comment:
Tom, what platform / STL are you building with? Perhaps on some STLs <string> does not include <limits>. We should probably spell it out specifically rather than rely on implementation details of the STL headers.